subprojects/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/CrossTaskIncrementalJavaCompilationIntegrationTest.groovy

```diff
@@ -72,6 +73,142 @@ abstract class CrossTaskIncrementalJavaCompilationIntegrationTest extends Abstra
         result.hasErrorOutput("package a is not visible")
     }
 
+    @Requires(TestPrecondition.JDK9_OR_LATER)
+    @Issue("https://github.com/gradle/gradle/issues/23067")
+    def "incrementally compilation works with modules with manual module path with #description"() {
+        file("impl/build.gradle") << """
+            def layout = project.layout
+            tasks.compileJava {
+                modularity.inferModulePath = false
+                options.compilerArgs.addAll($modulePathParameter)
+                doFirst {
+                    classpath = layout.files()
+                }
+            }
+        """
+        source api: ["package a; public class A {}"]
+        file("api/src/main/${language.name}/module-info.${language.name}").text = """
+            module api {
+                exports a;
+            }
+        """
+        source impl: [
+            "package b; import a.A; import c.C; public class B extends A {}",
+            "package c; public class C {}",
+            "package c.d; public class D {}"
+        ]
+        file("impl/src/main/${language.name}/module-info.${language.name}").text = """
+            module impl {
+                requires api;
+                exports b;
+                exports c;
+                exports c.d;
+            }
+        """
+        succeeds "impl:${language.compileTaskName}"
+
+        when:
+        impl.snapshot { source api: "package a; public class A { void m1() {} }" }
+
+        then:
+        succeeds "impl:${language.compileTaskName}", "--info"
+        impl.recompiledClasses("B", "module-info")
+
+        where:
+        description               | modulePathParameter
+        "--module-path=<modules>" | '["--module-path=\${classpath.join(File.pathSeparator)}"]'
+        "--module-path <modules>" | '["--module-path", "\${classpath.join(File.pathSeparator)}"]'
+        "-p <modules>"            | '["-p", "\${classpath.join(File.pathSeparator)}"]'
```

Comment:
This is because of a bug we had, where we didn't parse `--module-path=<modules>`  and `-p` correctly. And with that we snapshot module-path only when it is set in `--module-path <modules>` form. But this is actually a problem of `DefaultJavaCompileSpec`, so I will move test there.